File: 1824-minimum-sideway-jumps/1824-minimum-sideway-jumps.py
class Solution:
    def minSideJumps(self, obstacles: List[int]) -> int:
        dp=[[-1 for _ in range(3)] for _ in range(len(obstacles))]
        sys.setrecursionlimit(10**6)
        def solver(i,lane):
            if i >= len(obstacles) - 1:
                return 0
            if obstacles[i] == lane:
                return float("inf")
            if dp[i][lane - 1] != -1:
                return dp[i][lane - 1]
            if obstacles[i + 1] != lane:
                dp[i][lane - 1] = solver(i+1, lane)
            else:
                x, y = 0,0
                dp[i][lane - 1] = 1
                if lane == 1:
                    x, y = 2, 3
                elif lane == 2:
                    x, y = 1, 3
                else:
                    x, y = 1, 2
                dp[i][lane - 1] += min( solver(i, x), solver(i, y))
            return dp[i][lane-1]
        
        return solver(0,2)
                                       
                                     
        

File: 1824-minimum-sideway-jumps/1824-minimum-sideway-jumps.cpp
class Solution {
public:
    int solver(vector<int> &obs, int idx, int n, int lane, vector<vector<int>> &dp){
        if(idx == n - 1)
            return 0;
        if(obs[idx] == lane)
            return INT_MAX-100;
        if(dp[idx][lane - 1] != -1)
            return dp[idx][lane - 1];
        if(obs[idx+1]!=lane){
            dp[idx][lane - 1] = solver(obs, idx + 1, n, lane, dp);
        }
        else{
            int l1,l2;
            if(lane == 1){
                l1=2;
                l2=3;
            }
            else if(lane == 2){
                l1=1;
                l2=3;
            }
            else{
                l1=1;
                l2=2;
            }
            dp[idx][lane - 1] = 1;
            dp[idx][lane -1]+= min(solver(obs, idx ,n, l1,dp), solver(obs, idx, n , l2, dp));
        }
        return dp[idx][lane-1];
    }
    int minSideJumps(vector<int>& obstacles) {
        int n = obstacles.size();
        vector<vector<int>>dp(n, vector<int>(3,-1));
        return solver(obstacles, 0, n, 2, dp);
    }
};